File: Leetcode/0652. Find Duplicate Subtrees/0652.py
class Solution:
  def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
    ans = []
    count = Counter()

    def encode(root: Optional[TreeNode]) -> str:
      if not root:
        return ''

      left = encode(root.left)
      right = encode(root.right)
      encoding = str(root.val) + '#' + left + '#' + right

      if count[encoding] == 1:
        ans.append(root)
      count[encoding] += 1
      return encoding

    encode(root)
    return ans

File: Leetcode/0652. Find Duplicate Subtrees/0652.java
class Solutionclass Solution {
  public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
    List<TreeNode> ans = new ArrayList<>();
    Map<String, Integer> count = new HashMap<>();
    encode(root, count, ans);
    return ans;
  }

  private String encode(TreeNode root, Map<String, Integer> count, List<TreeNode> ans) {
    if (root == null)
      return "";

    final String left = encode(root.left, count, ans);
    final String right = encode(root.right, count, ans);
    final String encoding = root.val + "#" + left + "#" + right;

    if (count.getOrDefault(encoding, 0) == 1)
      ans.add(root);
    count.merge(encoding, 1, Integer::sum);
    return encoding;
  }
}
